src/WebJobs.Script/Host/ScriptHostManager.cs

```diff
@@ -55,31 +68,60 @@ public void RunAndBlock(CancellationToken cancellationToken = default(Cancellati
                 // signaled. That is fine - the restart will be processed immediately
                 // once we get to this line again. The important thing is that these
                 // restarts are only happening on a single thread.
-                _instance.RestartEvent.WaitOne();
+                newInstance.RestartEvent.WaitOne();
 
-                // stop the host fully
-                _instance.Stop();
-                _instance.Dispose();
+                // Orphan the current host instance. We're stopping it, so it won't listen for nay new functions
+                // it will finish any currently executing functions and then clean itself up.
+                // Spin around and create a new host instance.
+                Task tIgnore = Orphan(newInstance);
             }
             while (!_stopped);
         }
 
+        // Let the existing host instance finsih currently executing functions.
+        private async Task Orphan(ScriptHost instance)
+        {
+            await instance.StopAsync();
+            instance.Dispose();
+
+            lock (_liveInstances)
+            {
+                _liveInstances.Remove(instance);
+            }
+        }
+
         public void Stop()
         {
             _stopped = true;
 
             try
             {
-                if (_instance != null)
+                ScriptHost[] instances = GetLiveInstancesAndClear();
+
+                Task[] tasksStop = Array.ConvertAll(instances, instance => instance.StopAsync());
```

Comment:
If an instance is already stopped, I think calling StopAsync on it again will throw. Need to guard for that - I don't think the SDK is idempotent there.